File: 백준/Gold/20047. 동전 옮기기/동전 옮기기.py
import sys


def input():
    return sys.stdin.readline().rstrip()


def solution() -> str:
    n = int(input())
    S = list(input())
    T = list(input())
    i, j = map(int, input().split())

    twoCoin = ['_', '_']
    twoCoin[1] = S.pop(j)
    twoCoin[0] = S.pop(i)

    able = [True, True]
    for k in range(2):
        idx = 0
        si = 0
        ti = 0
        while ti < n and si < len(S):
            if S[si] != T[ti]:
                if idx < 2 and twoCoin[idx] == T[ti]:
                    idx += 1
                    si -= 1
                else:
                    able[k] = False
                    break
            ti += 1
            si += 1

        if able[k] and twoCoin[idx:] != T[ti:]:
            able[k] = False

        if k == 0:
            S.reverse()
            T.reverse()
            twoCoin.reverse()

    if able[0] or able[1]:
        return "YES"
    else:
        return "NO"


if __name__ == '__main__':
    print(solution())

File: 백준/Gold/20047. 동전 옮기기/동전 옮기기.cc
#include <bits/stdc++.h>
using namespace std;

int n;
string S, T;
vector<char> twoCoin(2);
int i, j;
string ans = "NO";

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n;
    cin >> S >> T;
    cin >> i >> j;

    twoCoin[0] = S[i];
    twoCoin[1] = S[j];
    S.erase(j, 1);
    S.erase(i, 1);

    for (int k = 0; k < 2; k++) {
        bool isAble = true;

        int si = 0, ti = 0, idx = 0;
        for (; si < S.size() && ti < T.size(); si++, ti++) {
            if (S[si] != T[ti]) {
                if (idx < twoCoin.size() && twoCoin[idx] == T[ti]) {
                    si--;
                    idx++;
                } else {
                    isAble = false;
                    break;
                }
            }
        }
        if (isAble) {
            for (; idx < twoCoin.size() && ti < T.size(); idx++, ti++) {
                if (twoCoin[idx] != T[ti]) {
                    isAble = false;
                }
            }
        }
        if (isAble) {
            ans = "YES";
            break;
        }
        reverse(S.begin(), S.end());
        reverse(T.begin(), T.end());
        reverse(twoCoin.begin(), twoCoin.end());
    }
    cout << ans;
    return 0;
}